### PR TITLE
feat(pt): Add Optimized DPA4 Neo inference path using CuTe

### PR DESCRIPTION
**Pull request by atulcthakur**
## Summary

This PR adds an opt-in, strict-FP32 CuTe DSL inference path for supported DPA4-Neo models.

The implementation accelerates SO(2) message passing and related operations used for energy, force, and stress inference. It includes architecture-aware dispatch for Ampere, Ada, Hopper, and Blackwell GPUs, automatic edge preparation, optional CuTe dependencies, and correctness and compile-integration tests.

## Usage

Install the optional dependencies:

```bash
pip install "deepmd-kit[torch,cute]"
```

Enable the complete Neo CuTe path before constructing or exporting the model:

```bash
export DP_NEO_CUTE_INFER=1
```

Existing DeepMD inference interfaces remain unchanged. Unsupported model configurations retain the existing implementation. `DP_NEO_CUTE_INFER` may coexist with `DP_TRITON_INFER=2`.

## Performance

For a strict-FP32, 3,072-atom water energy/force/stress workload, including neighbor-list construction, the CuTe path provides approximately:

- **1.25x–1.36x lower latency** than `torch.compile + DP_TRITON_INFER=2`
- **44–50% lower peak GPU allocation**, equivalent to a **1.8–2.0x smaller memory footprint**

The reported range covers measurements on NVIDIA A100, L40S, H100, and RTX PRO 6000 Blackwell GPUs.

## Correctness

Energy, force, atomic-energy, and virial outputs were compared against eager PyTorch. The optimized path passes `torch.allclose` with shared tolerances on the order of `5e-5`.

## Summary by CodeRabbit

* **New Features**
  * Added optional CuTe-accelerated SeZM inference for supported NVIDIA GPUs, including packed Wigner processing and optimized interaction, embedding, grid-product, and readout paths.
  * Added architecture-aware runtime selection, compilation caching, fallback behavior, and checkpoint-safe cache invalidation.
  * Added optional installation support and configuration documentation for CuTe inference.
* **Bug Fixes**
  * Inference now temporarily freezes parameters where required and reliably restores their original gradient settings.
* **Refactor**
  * Removed the experimental CuTe SO(2) value-path backend in favor of the newer CuTe inference implementation.

### CODE REVIEW

**Comment by atulcthakur** [comment]
Hi @OutisLi @njzjz 
Do you mind taking a look at this PR?

**Comment by OutisLi** [comment]
Thank you very much for this substantial contribution and for sharing the cross-architecture performance and memory results. The CuTe DSL work is valuable for DeePMD, and I appreciate the extensive correctness and integration coverage.

I am currently finishing #6001, which is a large DPA4-family refactor and performance series. It moves the shared SeZM/DPA4 model structure into `dpmodel`, aligns the PT and PT-expt implementations, reorganizes the kernel namespaces and dispatch boundaries, and changes both compiled training and PT2 inference/export paths, including Triton, CUDA, CuTe/cuTile integration, and dynamic-shape handling. This overlaps many of the SeZM descriptor, edge-cache, model, and export files changed here.

To avoid asking you to resolve a large conflict against code that is still changing, I think the cleanest sequence is to merge #6001 first, then rebase this PR onto the resulting structure and integrate the CuTe path through the shared DPA4-family dispatch. After that rebase, I will do a detailed review of the kernels and PT2/AOTI integration.

One integration direction I expect we should discuss after the rebase is using the generic `DP_CUTE_INFER` switch with shape-, layout-, dtype-, and device-capability-based dispatch, rather than encoding the `Neo` configuration in an environment-variable name. This lets the same backend extend to other DPA4-family shapes without adding model-size-specific flags. We should also verify that the selected path remains exportable as a PT2/AOTI package, because the LAMMPS inference path relies on PT2 artifacts.

Thank you again for the contribution and for your patience while #6001 lands.

**Comment by OutisLi** [comment]
One additional concern is PT2/AOTI compatibility, because DeepMD LAMMPS inference consumes frozen .pt2 packages.

My reading of the current integration is:

- DeepMD freeze traces the model on CPU and then moves the graph to the requested CUDA target, while the CuTe eligibility checks require CUDA tensors. Therefore the CuTe branch is not selected during tracing. A .pt2 export may still succeed, but that would only demonstrate that the fallback path was exported; it would not demonstrate that the CuTe kernels are present or used by the artifact.
- The K1 operators are registered as Python torch.library.custom_op implementations, and part of the execution state is addressed through Python-owned registries and integer handles. This can work for eager or torch.compile execution inside Python, but LAMMPS loads the AOTI artifact in a Python-less C++ process where those registrations and Python objects do not exist.
- A fake implementation enables tracing to reason about tensor metadata, but does not provide the runtime implementation required by the C++ AOTI runner. PyTorch explicitly recommends defining the operator and backend kernels in C++ for Python-less environments or AOTInductor: https://docs.pytorch.org/tutorials/advanced/python_custom_ops

CuTe DSL itself supports AOT compilation and C/C++ callable interfaces, so this is an integration gap rather than a fundamental CuTe limitation: https://docs.nvidia.com/cutlass/4.5.2/media/docs/pythonDSL/cute_dsl_general/dsl_ahead_of_time_compilation.html

For actual LAMMPS support, I think the CuTe artifacts need to be AOT-compiled behind a stable C++ operator ABI, or exposed through an Inductor lowering, then packaged and loaded with the DeepMD runtime. Frozen packed weights should also be explicit graph buffers/constants instead of process-local registry handles.

After rebasing onto #6001, could we please keep these as separate acceptance tests:

1. Freeze a supported K1 shape through the real DeepMD PT2 flow, load the resulting artifact, verify numerical parity, and prove that CuTe dispatch is used rather than the fallback path.
2. Load the same artifact through the Python-less C++ AOTI path used by LAMMPS and perform the same parity and dispatch checks.

If the initial scope is intentionally Python eager inference only, that limitation should be documented explicitly and the CuTe path should remain excluded from freeze until the C++ integration is available.

File: deepmd/pt_expt/kernels/cute/sezm/gie.py [hunk 1]
```diff
@@ -0,0 +1,740 @@
...
+            if node >= node_count:
+                load_node = 0
+            acc = _F32(0.0)
+            for idx in cutlass.range(channels + lane, full_width, _LANES, unroll=1):
+                acc += m_grad_out[load_node, idx].to(_F32) * m_out[load_node, idx].to(
+                    _F32
+                )
+            acc = cute.arch.warp_reduction_sum(acc)
+            if lane == 0 and node < node_count:
+                m_grad_inv[node, 0] = acc / m_inv[node, 0].to(_F32)
```
[the remark this answers is not in the document]
Comment (OutisLi) [reply deepmd/pt_expt/kernels/cute/sezm/gie.py:359]: Not applicable. `_launch_forward` is the implementation body of the registered `sezm_cute::gie_fused` custom op. The autograd-facing call receives the original tensors; `_gie_setup_context` saves them and `_gie_backward` returns gradients for `radial`, `zonal`, `inv_sqrt_deg`, and `gate`. The implementation-local `detach()` is only used for the raw DLPack/CuTe launch and does not sever the registered operator edge. `test_cuda_forward_and_wigner_radial_degree_gate_gradients` covers these gradients against the materialized reference.

File: deepmd/pt_expt/kernels/cute/sezm/so2/sm90/phase_c_attention_backward.py [hunk 2]
```diff
@@ -0,0 +1,842 @@
...
+        _require_tensor(name, tensor, shape, dtype, device)
+    if dst_ptr.numel() > 1:
+        destination_degrees = dst_ptr[1:] - dst_ptr[:-1]
+        torch._assert_async(
+            torch.all(
+                (destination_degrees >= 0) & (destination_degrees <= MAX_EDGES_PER_NODE)
+            ),
+            "SM90 Phase-C backward requires destination degrees in "
+            f"[0, {MAX_EDGES_PER_NODE}]",
+        )
```
[the remark this answers is not in the document]
Comment (OutisLi) [reply deepmd/pt_expt/kernels/cute/sezm/so2/sm90/phase_c_attention_backward.py:784]: Fixed in `2344b4dd`. The launch now asserts `dst_ptr[0] == 0` and `dst_ptr[-1] == edge_count` before checking per-node degree bounds.

File: deepmd/pt_expt/kernels/cute/sezm/output_grid/readout_l0.py [hunk 3]
```diff
@@ -0,0 +1,783 @@
...
+    ffn_in: torch.Tensor,
+) -> torch.Tensor | None:
+    """Return the optimized final `[N,32]` readout or ``None`` for fallback."""
+    if not runtime_policy.is_cute_infer_enabled():
+        return None
+    if not _inference_mode_is_frozen(output_ffn):
+        return None
+    if not _has_exact_neo_readout_contract(output_ffn, ffn_in):
+        return None
+    return _run_neo_readout_l0(output_ffn, ffn_in, readout_l0_product_cute)
```
[the remark this answers is not in the document]
Comment (OutisLi) [reply deepmd/pt_expt/kernels/cute/sezm/output_grid/readout_l0.py:643]: This is intentional fail-fast behavior. `DP_CUTE_INFER=1` is an explicit request for the optional CuTe runtime; silently falling back when that runtime is absent would report CuTe as enabled while not executing it. Structural or architecture ineligibility still returns `None` and falls back. `doc/install/easy-install.md` documents that missing runtime dependencies raise an error.

**Comment by atulcthakur** [comment]
Thanks @njzjz and @OutisLi 
I am aware of the C++ AOTI requirement for LAMMPS inference. As @OutisLi pointed out, CuTe supports AOT compilation. 
We can patch that in a follow up PR, along with potential improvements to the kernels.

**Review by OutisLi — APPROVED** [review]

**Comment by OutisLi** [comment]
> Thanks @njzjz and @OutisLi I am aware of the C++ AOTI requirement for LAMMPS inference. As @OutisLi pointed out, CuTe supports AOT compilation. We can patch that in a follow up PR, along with potential improvements to the kernels.

Thanks for your help with the CuTe realization, I already put this pr into the merge queue. Btw, I am wondering whether this kernel can be made generalized to support other sizes of DPA4? I believe this can be done since the handwritten c++ cuda code can realize this.

**Comment by atulcthakur** [comment]
Thanks @OutisLi 
Yes, kernels can indeed be generalized. We can discuss it over email.

File: deepmd/pt/model/descriptor/sezm_nn/block.py [hunk 4]
```diff
@@ -802,6 +812,30 @@ def _run_so2_unit_impl(
...
+                x,
+                edge_cache,
+                radial_feat,
+                dst_ptr=destination_row_ptr,
+                source_order=source_order,
+                source_ptr=source_row_ptr,
+            )
+            if accelerated is not None:
+                return accelerated
+            raise RuntimeError(
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/descriptor/sezm_nn/block.py]: When the packed Wigner cache is built, `_build_edge_wigner` returns `None, None, D_packed, edge_quat`, so the dense `D_full`/`Dt_full` operands are gone and this `raise` is the only remaining outcome if `maybe_run_cute_so2` declines. Eligibility is decided in two independent places: `is_packed_wigner_candidate` + `packed_wigner_edges_eligible` at cache-build time, and the dispatch-time contract checks inside `maybe_run_cute_so2` (`_register_cute_so2_state` can return `None` via `_module_state_is_aligned` even after a successful earlier registration). Any divergence, such as dtype after `edge_cache_to_dtype`, a shape the prepared runner rejects, or a compile failure in `_build_runner`, becomes a crash mid-inference rather than a slower correct run. That also contradicts `use_cute_infer()`'s docstring ("leaves unsupported shapes to the selected Triton, CUDA, or reference path") and `doc/install/easy-install.md`. Every other accelerated seam here (`make_*_value_path` returning `None`) keeps the reference path in charge. Could the packed cache retain the dense operands, or could the two eligibility predicates become one shared function so they cannot diverge? Same pattern in `deepmd/pt_expt/descriptor/dpa4_nn/block.py` around line 165.

File: deepmd/pt/model/descriptor/sezm.py [hunk 5]
```diff
@@ -1421,29 +1437,43 @@ def forward(
...
+                        edge_cache,
+                        n_nodes,
+                    )
+                )
                 with self._compute_mode_ctx(extended_coord.device):
-                    x = self._forward_blocks(x, edge_cache, rad_feat_per_block)
+                    x = self._forward_blocks(
+                        x,
+                        edge_cache,
+                        rad_feat_per_block,
```
Comment (wanghan-iapcm) [inline deepmd/pt/model/descriptor/sezm.py:1474]: This adds `comm_dict=comm_dict` to the nlist-path `_forward_blocks` call, which previously omitted it. On this path `n_nodes = nf * nloc` (`type_ebed` is built from `extended_atype[:, :nloc]`), so `x` has local rows only and ghosts go through `mapping`. `SeZMInteractionBlock._run_so2_unit` now reaches `exchange_ghost_features(x, comm_dict)`, whose contract is `(nall, D, 1, channels)` and which hands the buffer to `deepmd_export::border_op` to write rows `nlocal .. nlocal+nghost-1`, past the end of an `nloc`-row tensor. `forward_with_edges` has the explicit `parallel` guard, `nlocal` and `nf == 1` checks; this path has none, and the parameter is still documented as "(unused)" near line 1234. Today `SeZMModel` routes multi-rank inference through `forward_with_edges`, so this is latent, but it removes what kept the combination safe and nothing tests it. Either drop the argument again or add the same guard.

**Review by wanghan-iapcm — COMMENTED** [review]
I reviewed the integration surface (about 4.1k changed lines outside `deepmd/pt_expt/kernels/cute/`) rather than the kernel bodies. The default non-CuTe path looks unbroken: on this head `source/tests/consistent/descriptor/test_dpa4.py` (108 passed / 84 skipped), `source/tests/universal/dpmodel/descriptor/test_descriptor.py` (54 passed), `source/tests/pt_expt/descriptor/test_dpa4.py` (13 passed) and every new non-GPU test pass locally. Nothing here is blocking; two points are inline, the rest below.

**1. The prepared CuTe SO2 path drops the strict-FP32 and frozen-parameter gates that the fallback enforces.**
https://github.com/deepmodeling/deepmd-kit/blob/652e410f7d0955a7c95fa1dfd3a2827e57a84a25/deepmd/pt_expt/kernels/cute/sezm/so2/operation.py#L2034-L2079
`is_neo_so2_static_eligible` (https://github.com/deepmodeling/deepmd-kit/blob/652e410f7d0955a7c95fa1dfd3a2827e57a84a25/deepmd/pt_expt/kernels/cute/sezm/so2/operation.py#L383) requires `runtime_policy.uses_strict_fp32_matmul()`, `_module_uses_strict_fp32(block)` and `_module_is_frozen(block)`, but `_maybe_run_prepared_cute_so2` re-validates only `block.training`, device, `x.dtype`, autocast and the packed tensors. Once `prepare_cute_so2_blocks()` has stamped `_deepmd_cute_so2_state`, every later call takes the prepared branch. `uses_strict_fp32_matmul()` reads a mutable process global (`torch.backends.cuda.matmul.allow_tf32` / `fp32_precision`), and the runner still issues cuBLAS matmuls in `so2/linear.py` and `so2/structural_gate.py`, so a process that flips `allow_tf32` after registration silently gets TF32 under a contract advertised as strict FP32. `radial_phase_a.py` re-checks the same invariant on every call and raises, which suggests it is meant to hold per call. The same gap means a block whose parameters later get `requires_grad=True` receives no parameter gradients from this path.

**2. Two diverging implementations of the "SO2 thin wrapper" decision; the optimisation silently never engages on sm_86.**
https://github.com/deepmodeling/deepmd-kit/blob/652e410f7d0955a7c95fa1dfd3a2827e57a84a25/deepmd/pt/model/network/mlp.py#L51-L70
`runtime_policy.is_so2_thin_wrapper_enabled` is default-on for the whole SM80 profile (`SM80_PROFILE_CAPABILITIES = frozenset({(8, 0), (8, 6)})`, https://github.com/deepmodeling/deepmd-kit/blob/652e410f7d0955a7c95fa1dfd3a2827e57a84a25/deepmd/pt_expt/kernels/cute/sezm/runtime_policy.py#L19), and `sezm_model.py` uses that to stamp `_deepmd_cute_compile_visible_linear = True` on every `MLPLayer`. `MLPLayer.forward` then re-gates on `_use_so2_compile_visible_linear(xx.device)`, which re-parses the env from scratch and, absent an explicit override, requires capability `== (8, 0)` exactly. On sm_86 the SO2 dispatch takes the thin-wrapper topology while the MLP keeps `F.linear`. Not a correctness bug, but the two halves of one decision disagree. Worth collapsing to a single `runtime_policy` predicate.

**3. The GPU CI for this PR has not run.** `Test Python on CUDA` and `Test C++ on CUDA` both report `skipping` (run 35047886846) because the PR carries no `Test CUDA` label, so the new `cute` extra install, the `python -c "import cutlass.cute"` gate added in `.github/workflows/test_cuda.yml`, and every GPU-only test under `source/tests/pt/model/sezm_cute/` have not executed on this head. Locally on CPU that suite gives about 170 passed / 70 skipped, and the skipped set is exactly the kernel numerics. Could the label be applied and the job re-run before merge? Related: the `cute` extra pins `nvidia-cutlass-dsl[cu13]` while that workflow installs `.[gpu,test,lmp,cu12,cute,torch,jax]`; worth confirming the cu12/cu13 mix resolves on the runner, since the new import gate hard-fails rather than skipping.

**Smaller notes, not worth threads:**
- `_env_override` (https://github.com/deepmodeling/deepmd-kit/blob/652e410f7d0955a7c95fa1dfd3a2827e57a84a25/deepmd/pt_expt/kernels/cute/sezm/runtime_policy.py#L48) maps an unrecognised value to `False`, and callers read `is not False`, so `DP_CUTE_STRICT=2` silently disables the feature; every other `DP_*` gate in `deepmd/pt_expt/kernels/utils.py` raises `ValueError`.
- The new `deg_norm_floor <= 0.0` `ValueError` in `deepmd/pt/model/descriptor/sezm_nn/edge_cache.py` has no dpmodel counterpart, so the two backends now disagree on model validity; it is reachable on the default path for `version < 1.1` models where `deg_norm_floor` is the user-settable `eps`.
- `NvNeighborList._build_impl` (https://github.com/deepmodeling/deepmd-kit/blob/652e410f7d0955a7c95fa1dfd3a2827e57a84a25/deepmd/pt/utils/nv_nlist.py#L191) is a no-op split of `build()` with no other caller, and its docstring describes a "dispatch guard" that does not exist.
- The comment in `deepmd/dpmodel/descriptor/dpa4_nn/embedding.py` still says "the scatter of Step 5"; the new Step 3 pushed it to Step 6.

**Comment by njzjz** [comment]
Please fix conflicts
